### PR TITLE
[IN-1513] Run molecule tests as part of the CI

### PR DESCRIPTION
**Pull request by fehersanyi-bitrise**
### New Pull Request Checklist

*Please mark the points which you did / accept.*

- [ ] The tool I added is stable, and __does not require frequent updates__. _Preinstalled tools on the LTS stacks
  are not updated, so if the tool requires frequent updates you should handle the installation on-demand (see: [Install Any Additional Tool - DevCenter](https://bitrise-io.github.io/devcenter/tips-and-tricks/install-additional-tools/) for more information)_
- [ ] I added a version report line to [`system_report.sh`](/system_report.sh) for the new tool(s) I added
- [ ] I have added an entry to the CHANGELOG.md with a revisionID (the current date and the number of the change e.g.: 2019_10_11_1) and a brief description of the actual change
- [ ] I have updated the revisionID in [bitrise_profile](roles/profiles/files/bitrise_profile) **BITRISE_OSX_STACK_REV_ID** with the one I just created in the CHANGELOG.md

### CODE REVIEW

**Comment by MrBlaise** [comment]
It might be worth squashing some/most commits to make it easier to find the relevant ones in history after the merge.

**Comment by fehersanyi-bitrise** [comment]
@MrBlaise we squash [commits, and merge on big into master.

File: roles/ssh-setup/vars/debian.yml [hunk 1]
```diff
@@ -1,3 +1,3 @@
 ---
 home:
-  path: "/home"
+  path: "Users"
```
Comment (peterheja-bitrise) [inline roles/ssh-setup/vars/debian.yml:3]: This is a file loaded only for Debian-based Linuxes. This should stay as `/home`

File: _scripts/lead_time.sh [hunk 2]
```diff
@@ -3,4 +3,4 @@
 created_at=$(curl -s -H "Authorization: token $GITHUB_TOKEN" https://api.github.com/repos/bitrise-io/osx-box-bootstrap/pulls?state=closed | jq '.[0].created_at')
 envman add --key PR_CREATE_TIME --value "$created_at"
 closed_at=$(curl -s -H "Authorization: token $GITHUB_TOKEN" https://api.github.com/repos/bitrise-io/osx-box-bootstrap/pulls?state=closed | jq '.[0].closed_at')
-envman add --key PR_CLOSED_TIME --value "$closed_at"
\ No newline at end of file
+envman add --key PR_CLOSED_TIME --value "$closed_at"
```
Comment (peterheja-bitrise) [inline _scripts/lead_time.sh:6]: random whitespace change

File: roles/ssh-setup/tests/test_ssh.py [hunk 3]
```diff
@@ -1,13 +1,13 @@
 def test_ssh_config_exists(host):
-    home_path = "/home"
+    home_path = "Users"
     param_user = "linuxbrew"
```
Comment (peterheja-bitrise) [inline roles/ssh-setup/tests/test_ssh.py:3]: vagrant

File: roles/ssh-setup/tests/test_ssh.py [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
     ssh_config = host.file("/{}/{}/.ssh/config".format(home_path, param_user))
     assert ssh_config.exists
     assert ssh_config.contains("StrictHostKeyChecking")
     assert ssh_config.contains("UserKnownHostsFile")
 
 def test_ssh_folder(host):
-    home_path = "/home"
+    home_path = "Users"
     param_user = "linuxbrew"
```
Comment (peterheja-bitrise) [inline roles/ssh-setup/tests/test_ssh.py:11]: vagrant

File: roles/tests/molecule/ruby/converge.yml [hunk 5]
```diff
@@ -1,7 +1,6 @@
 ---
 - name: Converge
   hosts: all
-  tasks:
-    - name: "Include ruby"
-      include_role:
-        name: "ruby"
+  roles:
+    - role: ruby
+      param_user: "linuxbrew"
```
Comment (peterheja-bitrise) [inline roles/tests/molecule/ruby/converge.yml:6]: use vagrant user

File: _scripts/install_molecule_dependencies.sh [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    echo "Install apt dependencies..."
+    apt-get install -y build-essential libssl-dev libffi-dev python-dev &>"$BITRISE_DEPLOY_DIR"/apt_install.log
+
+    echo "Install pip3..."
+    apt install -y python3-pip &>/dev/null
+    pip install --upgrade pip &>/dev/null
+    pip_version="$(pip --version | awk '{print $1, $2}')"
+    echo "Installed $pip_version"
+
+    pip install virtualenv molecule testinfra docker pytest &>"$BITRISE_DEPLOY_DIR"/pip_install.log
```
Comment (MrBlaise) [inline _scripts/install_molecule_dependencies.sh:24]: virtualenv is installed here but not on MacOS, also I don't see if virtualenv is ever used

Comment (peterheja-bitrise) [inline _scripts/install_molecule_dependencies.sh]: virtualenv is unnecessary, if we don't use it actually

Comment (peterheja-bitrise) [reply _scripts/install_molecule_dependencies.sh]: also, in the future, it'd be nice to use a `requirements.txt` and a virtualenv, maybe.

Comment (fehersanyi-bitrise) [reply _scripts/install_molecule_dependencies.sh:24]: I added that because you guys said it, that is actually not a dependency 🧐 I'll remove it

Comment (MrBlaise) [reply _scripts/install_molecule_dependencies.sh]: On my part I meant creating a virtualenv and install stuff there, sorry for the confusion

File: _scripts/molecule_test.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+set -eo pipefail
+
+function run_test {
```
Comment (peterheja-bitrise) [inline _scripts/molecule_test.sh:4]: Why not `molecule test -s`? 
```
 Test (dependency, lint, cleanup, destroy, syntax, create, prepare,
  converge, idempotence, side_effect, verify, cleanup, destroy).
```

File: _scripts/molecule_test.sh [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+    molecule destroy -s "$1"
```
Comment (peterheja-bitrise) [inline _scripts/molecule_test.sh:5]: Will this actually work as intended? Will it really destroy an instance?
I don't see relevant molecule configuration changes, and I think these would run on the local docker

File: roles/ssh-setup/vars/debian.yml (hunk 1 above)
Comment (fehersanyi-bitrise) [reply roles/ssh-setup/vars/debian.yml:3]: nope, actually. this file loads for me locally, and the default.yml runs in the CI

File: roles/ssh-setup/tests/test_ssh.py (hunk 3 above)
Comment (fehersanyi-bitrise) [reply roles/ssh-setup/tests/test_ssh.py:3]: actually not, in the container it is linuxbrew by default

Comment (fehersanyi-bitrise) [reply roles/ssh-setup/tests/test_ssh.py:3]: but I can change that

File: _scripts/molecule_test.sh (hunk 7 above)
Comment (fehersanyi-bitrise) [reply _scripts/molecule_test.sh:4]: there were issues with the idenpotent part of `molecule test -s` a lot of shell modules are failing idempotency check as a hobby so I set it up according to the original guid of @lmesz-bitrise

File: _scripts/molecule_test.sh (hunk 8 above)
Comment (fehersanyi-bitrise) [reply _scripts/molecule_test.sh:5]: yes, I put that here so you can run it locally as well, on a clean container

File: _scripts/install_molecule_dependencies.sh (hunk 6 above)
Comment (fehersanyi-bitrise) [reply _scripts/install_molecule_dependencies.sh]: already removed that, requirements can be addedd later, the scope of the task is to run them as part of the CI without too many interference. If I implement every bit of best practice, I can work on this for a month :)

File: _scripts/molecule_test.sh (hunk 7 above)
Comment (peterheja-bitrise) [reply _scripts/molecule_test.sh:4]: okay

**Comment by fehersanyi-bitrise** [comment]
@peterheja-bitrise after your changes all the tests failed 🧐

**Comment by peterheja-bitrise** [comment]
fix the docker parts as well

File: roles/android/tasks/main.yml [hunk 9]
```diff
@@ -5,7 +5,7 @@
     path: "{{ android_ndk_home }}"
     state: directory
     owner: "{{ param_user }}"
-    mode: 0770
+    mode: "0770"
```
Comment (bitrise-tom) [inline roles/android/tasks/main.yml:8]: Why are all the permissions wrong in the first place?
Have we look upstream on out build pipeline to understand and fix this at install time?

Comment (fehersanyi-bitrise) [reply roles/android/tasks/main.yml:8]: I don't know, what would be the correct permission?

Comment (bitrise-tom) [reply roles/android/tasks/main.yml:8]: ah, apologies, it's a code correction, not new code. My bad

**Review by bitrise-tom — APPROVED** [review]

**Review by lmesz-bitrise — APPROVED** [review]

File: roles/ssh-setup/vars/debian.yml (hunk 1 above)
Comment (peterheja-bitrise) [reply roles/ssh-setup/vars/debian.yml:3]: https://github.com/bitrise-io/osx-box-bootstrap/blob/molecule-ci/roles/ssh-setup/tasks/main.yml#L2
```yaml
---
- name: load distribution specific variables
  include_vars: '{{ item }}'
  with_first_found:
    - files:
      - '{{ ansible_os_family }}.yml'
      - default.yml
```

this should make it only load `debian.yml` if your `ansible_os_family` is debian, which is true in the linuxbrew image, and on other linux distros, while it won't find the darwin/macos (whichever), therefore loads `default.yml`
if it is run in docker, that should be fine

**Review by peterheja-bitrise — APPROVED** [review]
lgtm

**Pull request merged**